AK/QuickSelect.h

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Random.h>
+#include <AK/StdLibExtras.h>
+
+namespace AK {
+// FIXME: Stole and adapted these two functions from `Userland/Demos/Tubes/Tubes.cpp` we really need something like this in `AK/Random.h`
+static inline double random_double()
+{
+    return get_random<u32>() / static_cast<double>(NumericLimits<u32>::max());
+}
+
+static inline int random_int(size_t min, size_t max)
+{
+    return min + round_to<size_t>(random_double() * (max - min));
+}
+
+// Implementations of common pivot functions
+namespace PivotFunctions {
+
+// Just use the first element of the range as the pivot
+// Mainly used to debug the quicksort algorithm
+// Good with random data since it has nearly no overhead
+// Attention: Turns the algorithm quadratic if used with already (partially) sorted data
+template<typename Collection, typename LessThan>
+size_t first_element([[maybe_unused]] Collection& collection, size_t left, [[maybe_unused]] size_t right, [[maybe_unused]] LessThan less_than)
+{
+    return left;
+}
+
+// Just use the middle element of the range as the pivot
+// This is what is used in AK::single_pivot_quick_sort in quicksort.h
+// Works fairly well with random Data
+// Works incredibly well with sorted data since the pivot is always a perfect split
+template<typename Collection, typename LessThan>
+size_t middle_element([[maybe_unused]] Collection& collection, size_t left, size_t right, [[maybe_unused]] LessThan less_than)
+{
+    return (left + right) / 2;
+}
+
+// Pick a random Pivot
+// This is the "Traditional" implementation of both quicksort and quick select
+// Performs fairly well both with random and sorted data
+template<typename Collection, typename LessThan>
+size_t random_element([[maybe_unused]] Collection& collection, size_t left, size_t right, [[maybe_unused]] LessThan less_than)
+{
+    return random_int(left, right);
+}
+
+// Implementation detail of median_of_medians
+// Whilst this looks quadratic in runtime, it always gets called with 5 or less elements so can be considered constant runtime
+template<typename Collection, typename LessThan>
+size_t partition5(Collection& collection, size_t left, size_t right, LessThan less_than)
+{
+    VERIFY((right - left) <= 5);
+    size_t i = left + 1;
+    while (i <= right) {
```

Comment:
Is there a particular reason that these aren't just `for`-loops?

Reply:
I don't think there's a good reason, I should definitely change this.